File: algorithms/187.Repeated_DNA_Sequences/Repeated_DNA_Sequences.py
class Solution:
    # @param s, a string
    # @return a list of strings
    def findRepeatedDnaSequences(self, s):
        value = {'A': 0, 'C': 1, 'G': 2, 'T': 3}
        appeared = set()
        used = set()
        ans = []

        if len(s) <= 10:
            return ans

        hash = 0
        for i in range(10):
            hash = (hash << 2) | value[s[i]]
        appeared.add(hash)

        for i in range(10, len(s)):
            hash = ((hash << 2) | value[s[i]]) & ((1<<20) - 1)
            if hash in used:
                continue
            if hash in appeared:
                ans.append(s[i - 9: i + 1])
                used.add(hash)
            else:
                appeared.add(hash)
        return ans

# debug
s = Solution()
print s.findRepeatedDnaSequences('AAAAAAAAAAAA')

class Solution:
    # @param s, a list of 1 length strings, e.g., s = ['h','e','l','l','o']
    # @return nothing
    def reverseWords(self, s):
        self.reverse(s, 0, len(s))
        
        i = 0
        for j in xrange(len(s) + 1):
            if j == len(s) or s[j] == ' ':
                self.reverse(s, i, j)
                i = j + 1
    
    def reverse(self, s, begin, end):
        for i in xrange((end - begin) / 2):
            s[begin + i], s[end - 1 - i] = s[end - 1 - i], s[begin + i]

if __name__ == '__main__':
    s = ['h','e','l','l','o', ' ', 'w', 'o', 'r', 'l', 'd']
    Solution().reverseWords(s)
    print s

File: algorithms/187.Repeated_DNA_Sequences/Repeated_DNA_Sequences.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Solution {

    private int get_value(char ch) {
        switch (ch) {
            case 'A': return 0;
            case 'C': return 1;
            case 'G': return 2;
            case 'T': return 3;
        }
        return -1;
    }

    public List<String> findRepeatedDnaSequences(String s) {
        List<String> ans = new ArrayList<String>();
        boolean appeared[] = new boolean[11000000];
        boolean used[] = new boolean[11000000];
        if (s.length() <= 10)
            return ans;

        int hash = 0;
        // get first substring's hash
        for (int i = 0; i < 10; i++)
            hash = (hash << 2) | get_value(s.charAt(i));
        appeared[hash] = true;

        // deal every substring
        for (int i = 10; i < s.length(); ++i) {
            hash = ((1<<20) - 1) & ((hash << 2) | get_value(s.charAt(i)));
            if (used[hash])
                continue;
            if (appeared[hash]) {
                ans.add(s.substring(i - 9, i + 1));
                used[hash] = true;
            } else
                appeared[hash] = true;
        }

        return ans;
    }

    // debug
    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        Solution s = new Solution();
        int[] input = {1, 2, 3, 1};
        System.out.println(s.findRepeatedDnaSequences("AAAAAAAAAAAA"));
    }
}

public class Solution {
    public int encode(String s) {
        int sum = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == 'A') {
                sum = sum * 4;
            } else if (s.charAt(i) == 'C') {
                sum = sum * 4 + 1;
            } else if (s.charAt(i) == 'G') {
                sum = sum * 4 + 2;
            } else {
                sum = sum * 4 + 3;
            }
        }
        return sum;
    }
    public List<String> findRepeatedDnaSequences(String s) {
        HashSet<Integer> hash = new HashSet<Integer>();
        HashSet<String> dna = new HashSet<String>();
        for (int i = 9; i < s.length(); i++) {
            String subString = s.substring(i - 9, i + 1);
            int encoded = encode(subString);
            if (hash.contains(encoded)) {
                dna.add(subString);
            } else {
                hash.add(encoded);
            }
        }
        List<String> result = new ArrayList<String>();
        for (String d: dna) {
            result.add(d);
        }
        return result;
    }
}
